[SE-0160] Limiting `@objc` inference

DougGregor opened the pull request:
This PR implements the majority of [SE-0160: Limiting `@objc` inference](https://github.com/apple/swift-evolution/blob/master/proposals/0160-objc-inference.md), which includes:

* An opt-in warning (enabled by `-warn-swift3-objc-inference`) that warns on every declaration that is implicitly `@objc` by the Swift 3 rules but would not be `@objc` under SE-0160, for use in conservative migration
* A warning about *uses* of the such declarations in Swift that involve the Objective-C runtime, e.g., `#selector`, `#keyPath`, calls via `AnyObject`, or overrides (if the inferred `@objc` is in a class extension)
* Introducing deprecated attributes into the Objective-C generated header for such declarations to catch any uses within Objective-C code for mixed-source projects
* Making `@IBInspectable` and `@GKInspectable` infer `@objc`.
* Compiler flags `-enable-swift3-objc-inference` and `-disable-swift3-objc-inference` to enable/disable the Swift 3 behavior. The former is the default in Swift 3 compatibility mode, the latter is the default in Swift 4
* Require `@objc` when `dynamic` is specified
* Introduce a new Swift runtime entry point `swift_objc_swift3ImplicitObjCEntrypoint` that is called whenever an Objective-C thunk created due to a deprecated `@objc` inference rule is called. Aside from being a hook one can latch onto in a debugger, one can set the environment variable `SWIFT_DEBUG_IMPLICIT_OBJC_ENTRYPOINT` to introduce logging: set it to 0 to disable logging, 1 to warn, 2 to warn with backtrace, and 3 to `abort()`.
* Add the attribute `@objcMembers`, which can be placed on a class to make it (and its subclasses) use the older `@objc` inference behavior.

There is still some work to make sure that `@objc` inference is working well for members that satisfy optional `@objc` protocol requirements for protocols that superclasses conform to, but that's separate.

Addresses rdar://problem/28497874, rdar://problem/31250936
Resolves #47058

DougGregor committed:
[SE-0160] Reorganize shouldMarkAsObjC() to separate out Swift 3 inference.

DougGregor committed:
[Type checker] Make the reason for @objc inference more descriptive.

Rather than the nebulous “do not diagnose”, separate out the two reasons for a non-diagnosed @objc: it’s a member of an Objective-C-derived subclass or it’s an accessor for a property.

DougGregor committed:
[SE-0160] Add an opt-in warning about deprecated @objc inference.

Introduce an opt-in warning (enabled by the frontend option
-warn-swift3-objc-inference) for each declaration for which @objc is
inferred based on Swift 3 rules that no longer apply after SE-0160.

DougGregor committed:
[SE-0160] Warn about uses of @objc declarations that used deprecated @objc inference.

When in Swift 3 compatibility mode without
`-warn-swift3-objc-inference`, warn on the *uses* of declarations that
depend on the Objective-C runtime that became `@objc` due to the
deprecated inference rule. This far more directly captures important
uses of the deprecated Objective-C entrypoints. We diagnose:

* `#selector` expressions that refer to one of these `@objc` members
* `#keyPath` expressions that refer to one of these `@objc` members
* Dynamic lookup (i.e., member access via `AnyObject`) that refers to
  one of these `@objc` members.

DougGregor committed:
[Serialization] Serialization for the @objc “Swift 3 inferred” bit

DougGregor committed:
Make Swift-3-inferred @objc explicit in test cases

DougGregor committed:
[SE-0160] Add “deprecated” attributes to the generated headers for Swift 3 inferred @objc.

When Swift 3 infers @objc using one of the rules deprecated in Swift 4, add a “deprecated” attribute to the declarations generated Objective-C header so that Objective-C gets warnings for uses of these APIs.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Still to do:

* Require explicit `@objc` to go along with `dynamic`
* Actually turn off this inference in Swift 4 mode, possibly behind a flag if we feel like staging it in more slowly
* Add a runtime entrypoint that gets called from the Objective-C thunks for these methods that can log/fail/etc. if they get called in practice
* (Separate) Improve the witness matching for protocols that come from superclasses so we don't miss obvious cases

DougGregor committed:
[SE-0160] Allow one to enable/disable Swift 3 @objc inference rules.

Introduce flags `-enable-swift3-objc-inference` and
`-disable-swift3-objc-inference` to enable/disable the Swift 3 `@objc`
inference rules. Under `-swift-version 3`, default to the former;
under `-swift-version 4`, default to the latter. For testing purposes,
one can provide either flag in eiher language mode.

DougGregor committed:
[SE-0160] @IBInspectable implies @objc

DougGregor committed:
[SE-0160] @GKInspectable implies @objc

DougGregor committed:
[SE-0160] Require '@objc' on 'dynamic' members in Swift 4.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor commented:
@swift-ci please smoke test Linux

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -2610,6 +2610,16 @@ ArtificialMainKind ClassDecl::getArtificialMainKind() const {
+bool ClassDecl::isXCTestCaseSubclass() const {
+  ASTContext &ctx = getASTContext();
+  for (auto D = getSuperclassDecl(); D; D = D->getSuperclassDecl()) {
+    if (D->getName() == ctx.Id_XCTestCase)
```
Worth also checking `hasClangDecl()`.

DougGregor replied:
We don't want to check `hasClangDecl()`, because corelibs XCTest also defines this, and we use this function to identify test cases there as well.

jrose-apple replied:
Hm. Maybe check that the containing module's name is "XCTest", then?

DougGregor replied:
Happy to do that, sure.

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -5062,6 +5072,45 @@ bool FuncDecl::isDeferBody() const {
+  if (retTy && !retTy->isVoid())
+    return false;
+
+  // 4. ...takes no parameters...
```
Do we really want to hardcode so many details? For IBAction it definitely bit us.

DougGregor replied:
This is the logic used by SourceKit; I'm okay with the details because they're already vetted as part of an existing contract.

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -5062,6 +5072,45 @@ bool FuncDecl::isDeferBody() const {
+       parentClass->getFormalAccess() < Accessibility::Internal))
+    return false;
+
+  // 6. ...and starts with "test".
```
Should this be a word-based starts-with test?

briancroom replied:
I think this is correct as-is. XCTest.framework includes `- (void)testfoo {}` / `func testfoo() {}` when it gathers test methods at runtime.

jrose-apple commented on include/swift/Option/Options.td:
```diff
@@ -247,6 +247,15 @@ def disable_swift_bridge_attr : Flag<["-"], "disable-swift-bridge-attr">,
+  Flags<[FrontendOption, HelpHidden]>,
+HelpText<"Enable Swift 3's @objc inference rules for NSObject-derived classes and 'dynamic' members (emulates Swift 3 behavior)">;
+
+def disable_swift3_objc_inference :
```
You made these driver options, but forgot to forward them in `addCommonFrontendArgs`. (If they're supposed to *just* be frontend options, please hide them in FrontendOptions.td.)

DougGregor replied:
Yikes, thanks! I should have written a driver test

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2285,11 +2286,13 @@ static Optional<ObjCReason> shouldMarkAsObjC(TypeChecker &TC,
     return ObjCReason::ExplicitlyIBOutlet;
   if (VD->getAttrs().hasAttribute<IBActionAttr>())
     return ObjCReason::ExplicitlyIBAction;
+  if (VD->getAttrs().hasAttribute<IBInspectableAttr>())
```
Huh, how did we miss this one? I guess cause there's no *inherent* reason for IBInspectable to imply "exposed to ObjC".

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -5103,8 +5103,9 @@ bool FuncDecl::isXCTestUnitTest() const {
-       parentClass->getFormalAccess() < Accessibility::Internal))
+      ((hasAccessibility() && getFormalAccess() < Accessibility::Internal) ||
+       (parentClass->hasAccessibility() &&
+        parentClass->getFormalAccess() < Accessibility::Internal)))
```
This is concerning. I would rather pass a LazyResolver here.

DougGregor replied:
Something is going horribly wrong on the Linux builders, which prompted this, I'll find the right thing to do here.

jrose-apple reviewed (commented):
Many comments. Each individual change seems pretty straightforward, though, which is good.

DougGregor commented on lib/Sema/CSApply.cpp:
```diff
@@ -910,6 +910,25 @@ namespace {
         closeExistential(ref, locator, /*force=*/openedExistential);
 
+        // If this attribute was inferred based on deprecated Swift 3 rules,
+        // complain.
```
I'm currently doing it across module boundaries, because I really did want to get the multiple-targets-in-a-project case. I don't think we have a way to tell the difference between a module that comes from a different package (that the user doesn't care about) vs. a different target (that the user does care about).

DougGregor commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -803,7 +810,33 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+
+  void printSwift3ObjCDeprecatedInference(ValueDecl *VD) {
+    auto attr = VD->getAttrs().getAttribute<ObjCAttr>();
+    if (!attr || !attr->isSwift3Inferred())
```
It'll end up tracking the language mode in which the declaration was compiled; we only mark `ObjCAttr` as "Swift 3 inferred" when Swift 3's `@objc` inference rules are applied.

DougGregor committed:
[Runtime] Add a “swift::warning” entrypoint for runtime warnings.

Like swift::fatalError(), but without the abort or crash logging.

DougGregor committed:
[SE-0160] Log uses of `@objc` thunks emitted due to deprecated @objc inference.

Introduce a new runtime entry point,
`swift_objc_swift3ImplicitObjCEntrypoint`, which is called from any
Objective-C method that was generated due to `@objc` inference rules
that were removed by SE-0160. Aside from being a central place where
users can set a breakpoint to catch when this occurs, this operation
provides logging capabilities that can be enabled by setting the
environment variable SWIFT_DEBUG_IMPLICIT_OBJC_ENTRYPOINT:

  SWIFT_DEBUG_IMPLICIT_OBJC_ENTRYPOINT=0 (default): do not log
  SWIFT_DEBUG_IMPLICIT_OBJC_ENTRYPOINT=1: log failed messages
  SWIFT_DEBUG_IMPLICIT_OBJC_ENTRYPOINT=2: log failed messages with
  backtrace
  SWIFT_DEBUG_IMPLICIT_OBJC_ENTRYPOINT=3: log failed messages with
  backtrace and abort the process.

The log messages look something like:

    ***Swift runtime: entrypoint -[t.MyClass foo] generated by
       implicit @objc inference is deprecated and will be removed in
       Swift 4

DougGregor commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2979,6 +2979,10 @@ ERROR(objc_operator, none,
       "@objc protocols may not have operator requirements", ())
+WARNING(objc_inference_swift3,none,
+        "inference of '@objc' for %select{'dynamic' members|members of "
+        "Objective-C--derived classes}0 is deprecated in Swift 4",
```
Double-hyphen is more correct, but fixed-width fonts stink. Fine :)

DougGregor commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2757,7 +2757,17 @@ void swift::markAsObjC(TypeChecker &TC, ValueDecl *D,
+    // Mark the attribute as having used Swift 3 inference, or create an
+    // implicit @objc for that purpose.
+    auto attr = D->getAttrs().getAttribute<ObjCAttr>();
+    if (!attr) {
```
Ah, you're right! I'd forgotten that we fully rely on the presence of `ObjCAttr`.

DougGregor commented on lib/Sema/TypeChecker.h:
```diff
@@ -515,24 +515,76 @@ withoutContext(TypeResolutionOptions options, bool preserveSIL = false) {
+  /// Is a witness to an @objc protocol requirement.
   WitnessToObjC,
+
+  // Note: Implicit reasons follow.
```
Yeah, it was a dumb comment. I'll remove it.

DougGregor committed:
Minor tweaks per Jordan's review (thanks!)

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Whoops. More SILGen tests need to get updated.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
@shahmishal  Look familiar?

 15:06:28 [xUnit] [ERROR] - Clock on this slave is out of sync with the master, and therefore 
15:06:28 I can't figure out what test results are new and what are old.
15:06:28 Please keep the slave clock in sync with the master.

DougGregor commented:
@swift-ci please test Linux

shahmishal commented:
```
00:00:05.478 Traceback (most recent call last):
00:00:05.478   File "/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/utils/update-checkout", line 26, in <module>
00:00:05.478     from SwiftBuildSupport import (
00:00:05.478   File "/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/utils/SwiftBuildSupport.py", line 26, in <module>
00:00:05.478     from swift_build_support import diagnostics  # noqa (E402)
00:00:05.478 ImportError: cannot import name diagnostics
00:00:05.484 Build step 'Execute shell' marked build as failure
```

DougGregor commented:
I'm actually interested in the full test, because I might have done something that behaves differently on iOS

shahmishal commented:
@swift-ci please test Linux

shahmishal commented:
```
Test Result (1 failure / +1)
Swift(iphonesimulator-i386).Interpreter/SDK.objc_swift3_deprecated_objc_inference.swift
```
https://ci.swift.org/job/swift-PR-osx/6511/

shahmishal commented:
@swift-ci clean test Linux

shahmishal commented:
@swift-ci clean test Linux

DougGregor committed:
[CMake] Warn about @objc inferred via the deprecated rules in stdlib/overlay builds

We want to know about these, both because they have an impact on our migration to Swift 4 and because we don’t want to vend extra Objective-C entry points.

DougGregor committed:
[Stdlib] Silence all warnings about deprecated @objc inference.

This eliminates 26 Objective-C entry points that we never intended to expose, but that came “for free” with the @objc inference rules.

DougGregor committed:
[SE-0160] Fix executable test to work with the iOS simulator

Provide SIMCTL_CHILD_ environment variables as well, which get passed along to the child of simctl. Additionally, use StdlibUnittest to handle the crash instead of `not —crash`, which doesn’t work through simctl.

DougGregor committed:
[Foundation] Address deprecated @objc inference warnings.

Another 15 unnecessary Objective-C entry points eliminated.

DougGregor committed:
[AppKit] Address deprecated @objc inference warnings.

DougGregor committed:
[SceneKit] Address deprecated @objc inference warnings.

DougGregor committed:
[Type checker] Remove unnecessary double-hyphen.

DougGregor commented:
Fixed the iOS issues and learned that I need the warning about deprecated `@objc` inference on the standard library and all of the overlays... which helped kill off ~47 unnecessary Objective-C entry points from the standard library and overlays.

jrose-apple commented on stdlib/public/SDK/AppKit/AppKit.swift:
```diff
@@ -85,6 +86,7 @@ extension NSImage : _ExpressibleByImageLiteral {
     self.init(named: name)
   }
 
+	@nonobjc
```
You're mixing tabs and spaces in some of these…

DougGregor committed:
[ObjectiveC] Resolve warning about deprecated @objc inference

DougGregor commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2716,6 +2741,24 @@ void swift::markAsObjC(TypeChecker &TC, ValueDecl *D,
+
+      TC.diagnose(D, diag::objc_inference_swift3, *deprecatedReason)
+        .highlight(dynamicLoc)
+        .fixItInsert(D->getAttributeInsertionLoc(false), "@objc ");
```
Sure

DougGregor committed:
[SE-0160] Address a few more comments from Jordan

DougGregor committed:
[Driver] Surface -warn-swift3-objc-inference to swiftc.

The `-warn-swift3-objc-inference` option turns out to be extremely
useful in vetting code for unintended `@objc` entry points, so make it
available directly on `swiftc`.

But, bury the enable/disable flags under `-frontend` (they were
effectively there anyway because the driver wasn't propagating them).

DougGregor committed:
[SE-0160] Improve diagnostics for -warn-swift3-objc-inference.

Split the warning into two warnings, because we want different Fix-It behavior.

* For the ‘dynamic’ warning, drop the “in Swift 4” part and keep the @objc Fix-It on the main warning. This is always auto-applyable.

* For the inferred-for-members-of-Objective-C-derived-classes warning, drop the “in Swift 4” part. More importantly, split off two notes: one with the @objc Fix-It (Swift 3 behavior) and one with the @nonobjc Fix-It (Swift 4 behavior). This makes it clearer that there is a choice, and neither is “obviously” correct.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please smoke test OS X

DougGregor commented:
The swiftpm thing was separate and should be fixed now.

DougGregor commented:
@swift-ci please test OS X

DougGregor committed:
[SE-0160] Don't generate runtime warnings for deprecated @objc inference of 'dynamic'.

When a method is dynamic, we always call through the Objective-C
runtime, which leads to a huge number of false positives. Suppress the
runtime calls here.

DougGregor committed:
[SE-0160] Diagnose overrides of members in extensions using deprecated @objc.

Overriding of members introduced in class extensions depends on the
presence of an Objective-C entrypoint. When we override such a
member---which used the deprecated @objc inference rule and occurs in
a class extension, where non-@objc methods currently cannot be
overridden---warn about the use of explicit @objc.

DougGregor committed:
[SE-0160] Don't model explicit 'dynamic' as deprecated @objc inference.

When the 'dynamic' modifier is explicitly written on a declaration, it
requires the Objective-C runtime. In Swift 3, this (directly) implies
'@objc'. In Swift 4, it requires that the '@objc' be written
explicitly. Hence, a Swift 3 -> Swift 4 migration will always retain
the @objc-ness of the entry point, so we don't need to treat it as
"inferred".

DougGregor committed:
[SE-0160] Add a test for #keyPath warnings about inferred @objc

DougGregor committed:
[SE-0160] Warn var/subscript using deprecated @objc and explicit accessor @objc

Warn about cases where a storage declaration (property or subscript)
has an accessor with an explicit @objc, but for which the storage
declaration itself is only @objc due to deprecated @objc inference.

DougGregor committed:
[SE-0160] Add @objcMembers attribute and infer it on XCTestCase.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge
